source/ceylon/buffer/codec/IncrementalCodec.ceylon

```diff
@@ -33,14 +34,27 @@ shared class ChunkConvert<ToMutable, FromImmutable, ToSingle, FromSingle>(conver
             return 0;
         }
         
-        // Don't use each() as we might need to stop before all of input is read
         variable Integer readCount = 0;
-        for (inputElement in input) {
-            readCount++;
-            value elements = pieceConverter.more(inputElement).iterator();
+        while (true) {
+            Iterator<ToSingle> elements;
+            value inputElement = provider();
+            if (is Finished inputElement) {
+                elements = pieceConverter.done().iterator();
+            }
+            else {
+                readCount++;
+                elements = pieceConverter.more(inputElement).iterator();
+            }
             while (is ToSingle element = elements.next()) {
```

Comment:
The compiler really can't figure out those two cases are semantically equivalent? If it can't, it should be fixed. I prefer writing logical statements in the simplest terms, rather than a double negative.